database/download.go

```diff
@@ -0,0 +1,87 @@
+package database
+
+import (
+	"context"
+	"time"
+
+	"gitlab.com/NebulousLabs/errors"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+)
+
+// Download ...
+type Download struct {
+	ID        primitive.ObjectID  `bson:"_id,omitempty" json:"_id"`
+	UserID    primitive.ObjectID  `bson:"user_id,omitempty" json:"user_id"`
+	SkylinkID primitive.ObjectID  `bson:"skylink_id,omitempty" json:"skylink_id"`
+	Timestamp primitive.Timestamp `bson:"timestamp" json:"timestamp"`
+}
+
+// DownloadByID fetches a single download from the DB.
+func (db *DB) DownloadByID(ctx context.Context, id primitive.ObjectID) (*Download, error) {
+	var d Download
+	filter := bson.D{{"_id", id}}
+	sr := db.staticDownloads.FindOne(ctx, filter)
+	err := sr.Decode(&d)
+	if err != nil {
+		return nil, err
+	}
+	return &d, nil
+}
+
+// DownloadCreate registers a new download.
+func (db *DB) DownloadCreate(ctx context.Context, user User, skylink Skylink) (*Download, error) {
+	if user.ID.IsZero() {
+		return nil, errors.New("invalid user")
+	}
+	if skylink.ID.IsZero() {
+		return nil, errors.New("invalid skylink")
+	}
+	up := Download{
+		UserID:    user.ID,
+		SkylinkID: skylink.ID,
+		Timestamp: primitive.Timestamp{T: uint32(time.Now().Unix())},
+	}
+	ior, err := db.staticDownloads.InsertOne(ctx, up)
+	if err != nil {
+		return nil, err
+	}
+	up.ID = ior.InsertedID.(primitive.ObjectID)
+	return &up, nil
+}
+
+// DownloadsBySkylink fetches all downloads of this skylink
+func (db *DB) DownloadsBySkylink(ctx context.Context, skylink Skylink) ([]Download, error) {
+	if skylink.ID.IsZero() {
+		return nil, errors.New("invalid skylink")
+	}
+	filter := bson.D{{"skylink_id", skylink.ID}}
+	c, err := db.staticDownloads.Find(ctx, filter)
+	if err != nil {
+		return nil, err
+	}
+	downloads := make([]Download, 0)
+	err = c.All(ctx, &downloads)
+	if err != nil {
+		return nil, err
+	}
+	return downloads, nil
+}
+
+// DownloadsByUser fetches all downloads by this user
+func (db *DB) DownloadsByUser(ctx context.Context, user User) ([]Download, error) {
+	if user.ID.IsZero() {
+		return nil, errors.New("invalid user")
+	}
+	filter := bson.D{{"user_id", user.ID}}
+	c, err := db.staticDownloads.Find(ctx, filter)
```

Comment:
Adding pagination should be as easy as to use `FindAndCount` here with offset and skip passed in as options. Not sure if you'll have `FindAndCount` at your disposal but most libraries have some sort of built in query to support list routes.

Reply:
Pagination added.